docs/proposals/memory-management.md

```diff
@@ -0,0 +1,165 @@
+
+   c. `Best-Effort`: Memory is never guaranteed. This class of containers are the first ones to be killed by the kernel whenever there is memory pressure on the node. 
+
+6. Detect OOM kills on containers: Docker tracks OOM kills on its containers. Kubelet exposes this information via [ContainerStatus.State](../pkg/api/v1/types.go#L1033), where the state is set to `ContainerStateTerminated` and `Reason` is set to `OOM Killed`.
+
+#### What are some of the known limitations?
+
+1. The scheduler does not take into account the total usage on the node. Hence, memory can be easily over-committed. Whenever a job is scheduled without resource requests, the node becomes over-committed. 
+
+2. Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+
+3. A container in a pod with restart policy set to `Always` will be restarted continuously even if it OOMs. 
+
+4. Other than the Guaranteed class, containers in other classes don't have a memory limit. A `Best-effort` container can use a lot of memory and result in high tail latencies for `Guaranteed` and `Burstable` class containers.
+
+5. Kubelet does not act on OOM kills. A container can choose to continue running when one or more processes in the container were killed by the kernel OOM killer. 
+
+6. In the event of an OOM kill due to system memory pressure, the kubelet cannot identify the container of the killed process.
+
+7. Memory usage of system daemons is unrestricted. This results in varying amount of memory availability on the node.
+
+8. All burstable containers that exceed their request are equally likely to be killed. Ideally, the job that exceeds the most should be killed. 
+  
+##### How to address these limitations?
+
+We can improve memory management by making use of [kernel memcg features](https://www.kernel.org/doc/Documentation/cgroups/memory.txt), and by modifying the current system behavior. Some of these improvements are required in the short-term. The rest are either complex or require significant changes to the system.
+
+###### What can be done right away?
+
+1. Cleanup the entire container, whenever a OOM kill is detected in a container. Users should be able to override the default behavior for containers that have an in-built process manager that can handle OOM kills.
+
+2. Set memory soft limits on Burstable containers to their request. This should result in "Burstable" containers being pushed to their requested memory limit under system memory pressure. 
+
+3. Restrict resource usage of system processes. Since the clusters are heterogeneous, the kubelet will have to come up with a limit for system processes dynamically, and apply those limits. 
```

Comment:
Docker's memory usage profile is not deterministic. We have usage as high as a GB and it is most likely proportional to the number of containers (and execs), which is usually proportional to the size of the machine.